### PR TITLE
refactor(ivy): simplify `walkTNodeTree` method for readability

### PR DESCRIPTION
**Pull request by mhevery**

### CODE REVIEW

File: packages/core/src/render3/node_manipulation.ts [hunk 1]
```diff
@@ -647,7 +512,7 @@ function nativeAppendChild(renderer: Renderer3, parent: RElement, child: RNode):
 
 function nativeAppendOrInsertBefore(
     renderer: Renderer3, parent: RElement, child: RNode, beforeNode: RNode | null) {
-  if (beforeNode) {
+  if (beforeNode != null) {
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/node_manipulation.ts]: `!==` ?

File: packages/core/src/render3/node_manipulation.ts [hunk 2]
```diff
@@ -843,3 +709,159 @@ function appendProjectedNode(
...
+  }
+  for (let i = CONTAINER_HEADER_OFFSET; i < lContainer.length; i++) {
+    const lView = lContainer[i] as LView;
+    applyView(renderer, action, lView, renderParent, anchor);
+  }
+}
+
+
+/**
+ * `applyContainer` performs operation on the container and its views as specified by `action`
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/node_manipulation.ts]: Looks like comment here was copied from the "container" while here we are dealing with "element container".

File: packages/core/src/render3/node_manipulation.ts [hunk 3]
```diff
@@ -843,3 +709,159 @@ function appendProjectedNode(
...
+ * (insert, detach, destroy)
+ *
+ * Inserting a Container is complicated by the fact that the container may have Views which
+ * themselves
+ * have container or projections.
+ *
+ * @param renderer Render to use
+ * @param action action to perform (insert, detach, destroy)
+ * @param lView The LView which needs to be inserted, detached, destroyed.
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/node_manipulation.ts]: `tElementContainerNode` is not part of the argument list

File: packages/core/src/render3/node_manipulation.ts [hunk 4]
```diff
@@ -60,177 +62,41 @@ const enum WalkTNodeTreeAction {
...
-      // Otherwise, this is a View
-      nextTNode = tNode.child;
+function executeNodeAction(
+    action: WalkTNodeTreeAction, renderer: Renderer3, parent: RElement | null,
+    nodeOrLContainer: RNode | LContainer | LView, beforeNode?: RNode | null) {
+  ngDevMode && assertDefined(nodeOrLContainer, '\'nodeOrLContainer\' is undefined');
+  let wrappedNode: any = nodeOrLContainer;
+  let lContainer: LContainer|undefined;
+  let isComponent = false;
+  while (Array.isArray(wrappedNode)) {
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/node_manipulation.ts]: It wasn't clear to me while reading the code what is going on here. Add comments explaining what and why we are doing here?

File: packages/core/src/render3/node_manipulation.ts [hunk 5]
```diff
@@ -60,177 +62,41 @@ const enum WalkTNodeTreeAction {
...
-          currentView = componentView[PARENT] !as LView;
-          nextTNode = currentView[TVIEW].data[head.index] as TNode;
-        }
-      }
-    } else {
-      // Otherwise, this is a View
-      nextTNode = tNode.child;
+function executeNodeAction(
+    action: WalkTNodeTreeAction, renderer: Renderer3, parent: RElement | null,
+    nodeOrLContainer: RNode | LContainer | LView, beforeNode?: RNode | null) {
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/node_manipulation.ts]: `nodeOrLContainer` variable name is misleading as it could also be a `LView`

File: packages/core/src/render3/node_manipulation.ts [hunk 6]
```diff
@@ -253,14 +119,13 @@ export function addRemoveViewFromContainer(
     viewToWalk: LView, insertMode: true, beforeNode: RNode | null): void;
 export function addRemoveViewFromContainer(viewToWalk: LView, insertMode: false): void;
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/node_manipulation.ts]: Please rename `viewToWalk` -> `lView` in all overloaded version of this function + documentation

File: packages/core/src/render3/node_manipulation.ts [hunk 7]
```diff
@@ -60,177 +62,41 @@ const enum WalkTNodeTreeAction {
...
-        }
-      }
+    if (isLView(wrappedNode)) {
+      isComponent = true;
     }
-    tNode = nextTNode;
+    wrappedNode = wrappedNode[0];
   }
-}
+  const node: RNode = wrappedNode;
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/node_manipulation.ts]: `node` -> `rNode`?

File: packages/core/src/render3/node_manipulation.ts [hunk 8]
```diff
@@ -60,177 +62,41 @@ const enum WalkTNodeTreeAction {
...
-          }
-        } else {
-          nextTNode = tNode.next;
-        }
-      }
+    if (isLView(wrappedNode)) {
+      isComponent = true;
     }
-    tNode = nextTNode;
+    wrappedNode = wrappedNode[0];
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/node_manipulation.ts]: `0` is a "magic number" here, did you mean `HOST`?

File: packages/core/src/render3/node_manipulation.ts [hunk 9]
```diff
@@ -843,3 +709,159 @@ function appendProjectedNode(
...
+    const viewRootTNodeType = viewRootTNode.type;
+    if (viewRootTNodeType === TNodeType.ElementContainer) {
+      applyElementContainer(
+          renderer, action, lView, viewRootTNode as TElementContainerNode, renderParent,
+          beforeNode);
+    } else if (viewRootTNodeType === TNodeType.Projection) {
+      applyProjection(
+          renderer, action, lView, viewRootTNode as TProjectionNode, renderParent, beforeNode);
+    } else {
+      ngDevMode && assertNotEqual(viewRootTNodeType, TNodeType.View, 'Unexpected TNodeType.View');
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/node_manipulation.ts]: Maybe turn those assertions into a "whitelist" instead (at this point it can be only a container or an element, right?

**Review by pkozlowski-opensource — CHANGES_REQUESTED** [review]
Left some in-line comments (mostly nits). Love the direction, generally speaking.

The are material tests failures, they look legit.

Also, wasn't clear for me if this is still WIP (there are some comments left in the code that either needs to be addressed and / or turned into TODOs / JIRA).

Finally, it needs golden file update.

File: packages/core/src/render3/node_manipulation.ts (hunk 5 above)
Comment (mhevery) [reply packages/core/src/render3/node_manipulation.ts]: renamed

File: packages/core/src/render3/node_manipulation.ts (hunk 4 above)
Comment (mhevery) [reply packages/core/src/render3/node_manipulation.ts]: added a comment

File: packages/core/src/render3/node_manipulation.ts (hunk 8 above)
Comment (mhevery) [reply packages/core/src/render3/node_manipulation.ts]: fixed

File: packages/core/src/render3/node_manipulation.ts (hunk 7 above)
Comment (mhevery) [reply packages/core/src/render3/node_manipulation.ts]: fixed

File: packages/core/src/render3/node_manipulation.ts (hunk 6 above)
Comment (mhevery) [reply packages/core/src/render3/node_manipulation.ts]: fixed

File: packages/core/src/render3/node_manipulation.ts (hunk 1 above)
Comment (mhevery) [reply packages/core/src/render3/node_manipulation.ts]: I specifically want to check for `null` and `undefined`, left a comment

File: packages/core/src/render3/node_manipulation.ts (hunk 9 above)
Comment (mhevery) [reply packages/core/src/render3/node_manipulation.ts]: fixed

File: packages/core/src/render3/node_manipulation.ts (hunk 2 above)
Comment (mhevery) [reply packages/core/src/render3/node_manipulation.ts]: fixed

File: packages/core/src/render3/node_manipulation.ts (hunk 3 above)
Comment (mhevery) [reply packages/core/src/render3/node_manipulation.ts]: fixed

File: packages/core/src/render3/node_manipulation.ts [hunk 10]
```diff
@@ -843,3 +712,156 @@ function appendProjectedNode(
...
+    for (let i = 0; i < nodeToProject.length; i++) {
+      const node = nodeToProject[i];
+      executeNodeAction(action, renderer, renderParent, node, beforeNode);
+    }
+  } else {
+    let projectionTNode: TNode|null = nodeToProject;
+    const projectedComponentLView = findComponentView(componentLView[PARENT] as LView);
+    while (projectionTNode !== null) {
+      const node = projectedComponentLView[projectionTNode.index];
+      executeNodeAction(action, renderer, renderParent, node, beforeNode);
```
Comment (kara) [inline packages/core/src/render3/node_manipulation.ts]: What if a projected node is an ng-container? Doesn't seem like we handle that case?

**Review by kara — CHANGES_REQUESTED** [review]
Yay, already looks so much easier to follow!  Some comments below

File: packages/core/src/render3/node_manipulation.ts [hunk 11]
```diff
@@ -843,3 +712,156 @@ function appendProjectedNode(
...
+ *
+ * Things get more complicated with containers and projections. That is because coming across:
+ * - Container: implies that we have to insert/remove/destroy the views of that container as well
+ *              which in turn can have their own Containers at the View roots.
+ * - Projection: implies that we have to insert/remove/destroy the nodes of the projection. The
+ *               complication is that the nodes we are projecting can themselves have Containers
+ *               or other Projections.
+ *
+ * As you can see this is a very recursive problem. Yes recursion is not most efficient but the
+ * code is complicated enough that trying to implemented with recursion becomes unmaintainable.
```
[the remark this answers is not in the document]
Comment (pkozlowski-opensource) [reply packages/core/src/render3/node_manipulation.ts]: Reworded this part of the comment.

File: packages/core/src/render3/node_manipulation.ts [hunk 12]
```diff
@@ -843,3 +712,156 @@ function appendProjectedNode(
...
+    renderer: Renderer3, action: WalkTNodeTreeAction, lContainer: LContainer,
+    renderParent: RElement | null, beforeNode: RNode | null | undefined) {
+  ngDevMode && assertLContainer(lContainer);
+  const anchor = lContainer[NATIVE];  // LContainer has its own before node.
+  const native = unwrapRNode(lContainer);
+  if (anchor !== native) {
+    // This is very strange to me (Misko). I would expect that the native is same as anchor. I don't
+    // see a reason why they should be different, but they are.
+    //
+    // If they are we need to process the second anchor as well.
```
[the remark this answers is not in the document]
Comment (pkozlowski-opensource) [reply packages/core/src/render3/node_manipulation.ts]: I think that this is generally the case when asking for a `ViewContainerRef` on any element other than `<ng-template>` / `<ng-container>` (those are already represented as comment nodes in the DOM). This is _exactly_ what I was getting at [in my doc](https://docs.google.com/document/d/1X5hH828uVmtJWLyi7txmyi98ETGEstZf8ANDOYajMi4/edit#heading=h.m0c2g3ussi36).

File: packages/core/src/render3/node_manipulation.ts (hunk 10 above)
Comment (pkozlowski-opensource) [reply packages/core/src/render3/node_manipulation.ts]: Yep, this is a missing case for which we don't have any unit test. Adding a test and fixing.

File: packages/core/src/render3/node_manipulation.ts [hunk 13]
```diff
@@ -843,3 +712,156 @@ function appendProjectedNode(
...
+  while (elementContainerRootTNode) {
+    const elementContainerRootTNodeType = elementContainerRootTNode.type;
+    const nodeOrLContainer = lView[elementContainerRootTNode.index];
+    if (elementContainerRootTNodeType === TNodeType.ElementContainer) {
+      applyElementContainer(
+          renderer, action, lView, elementContainerRootTNode as TElementContainerNode, renderParent,
+          beforeNode);
+    } else {
+      ngDevMode && assertNodeOfPossibleTypes(
+                       elementContainerRootTNode, TNodeType.Element, TNodeType.Container);
```
[the remark this answers is not in the document]
Comment (pkozlowski-opensource) [reply packages/core/src/render3/node_manipulation.ts]: Yep, this is a missing case for which we don't seem to have any test... Adding a test and fixing.

File: packages/core/src/render3/node_manipulation.ts (hunk 10 above)
Comment (pkozlowski-opensource) [reply packages/core/src/render3/node_manipulation.ts]: On top of this we are missing a case here of a projection node being part of projection (re-projection case). Adding a test and fix for this one as well.

File: packages/core/src/render3/node_manipulation.ts [hunk 14]
```diff
@@ -843,3 +712,156 @@ function appendProjectedNode(
     }
   }
 }
+
+
+/**
+ * `applyView` performs operation on the view as specified in `action` (insert, detach, destroy)
```
Comment (jelbourn) [inline packages/core/src/render3/node_manipulation.ts]: Function descriptions are typically written like:
"Performs the specified `action` (insert, detach, destroy) on the given view."

(similar for other functions)

File: packages/core/src/render3/node_manipulation.ts (hunk 11 above)
Comment (jelbourn) [inline packages/core/src/render3/node_manipulation.ts]: nit: this comment block is more conversational than authoritative, I would reword a bit like
```
For insertion *without* either projection or containers, we simply iterate over the root nodes
and perform the `action` on each.

With containers or projection, we're dealing with constructs that have their _own_
containers or further projection. While a recursion solution incurs some performance
trade-offs [citation needed], this approach makes the code significantly more maintainable.
```

File: packages/core/src/render3/node_manipulation.ts [hunk 15]
```diff
@@ -843,3 +712,156 @@ function appendProjectedNode(
...
+ * @param lView The LView which needs to be inserted, detached, destroyed.
+ * @param renderParent parent DOM element for insertion/removal.
+ * @param beforeNode Before which node the insertions should happen.
+ */
+function applyView(
+    renderer: Renderer3, action: WalkTNodeTreeAction, lView: LView, renderParent: RElement | null,
+    beforeNode: RNode | null | undefined) {
+  const tView = lView[TVIEW];
+  ngDevMode && assertNodeType(tView.node !, TNodeType.View);
+  let viewRootTNode: TNode|null = tView.node !.child;  // Why is this not `tView.firstChild`???;
```
Comment (jelbourn) [inline packages/core/src/render3/node_manipulation.ts]: Resolve this comment?

File: packages/core/src/render3/node_manipulation.ts [hunk 16]
```diff
@@ -843,3 +712,156 @@ function appendProjectedNode(
...
+ * As you can see this is a very recursive problem. Yes recursion is not most efficient but the
+ * code is complicated enough that trying to implemented with recursion becomes unmaintainable.
+ *
+ * @param renderer Render to use
+ * @param action action to perform (insert, detach, destroy)
+ * @param lView The LView which needs to be inserted, detached, destroyed.
+ * @param renderParent parent DOM element for insertion/removal.
+ * @param beforeNode Before which node the insertions should happen.
+ */
+function applyView(
```
Comment (jelbourn) [inline packages/core/src/render3/node_manipulation.ts]: `applyViewAction` or `appyActionToView`?

**Comment by pkozlowski-opensource** [comment]
Closing in favour of #31182
